account-compression/programs/account-compression/src/lib.rs

```diff
@@ -173,69 +202,137 @@ pub mod spl_account_compression {
-    //         proof.push(node.key().to_bytes());
-    //     }
-    //     fill_in_proof_from_canopy(canopy_bytes, header.max_depth, index, &mut proof)?;
-    //     assert_eq!(proof.len(), max_depth as usize);
-
-    //     let id = ctx.accounts.merkle_tree.key();
-    //     // A call is made to ConcurrentMerkleTree::initialize_with_root(root, leaf, proof, index)
-    //     let change_log = merkle_tree_apply_fn!(
-    //         header,
-    //         id,
-    //         tree_bytes,
-    //         initialize_with_root,
-    //         root,
-    //         leaf,
-    //         &proof,
-    //         index
-    //     )?;
-    //     wrap_event(change_log.try_to_vec()?, &ctx.accounts.log_wrapper)?;
-    //     update_canopy(canopy_bytes, header.max_depth, Some(change_log))
-    // }
+    /// For extremley large trees that cannot be initialized with root in a single transaction,
+    /// this allows a user to fill a buffer of proofs that can be used to initialize the tree with roots.
+    pub fn fill_proof_buffer(
+        ctx: Context<FillProofBuffer>,
+        max_depth: u32,
+        partial_proof: Vec<[u8; 32]>,
+        index: u32,
+    ) -> Result<()> {
+        let buffer: &mut [u8] = &mut *ctx.accounts.proof_buffer.try_borrow_mut_data()?;
+        let len = buffer.len();
+        let owner = ctx.accounts.proof_buffer.owner;
+        let (max, remaining) = if len == 0 {
+            if index != 0 {
+                return Err(AccountCompressionError::ProofIndexOutOfBounds.into());
+            }
+            if owner != &system_program::id() {
+                return Err(AccountCompressionError::IncorrectAccountOwner.into());
+            }
+            let space = (max_depth * 32) + 4;
+            let lamports = Rent::get()?.minimum_balance(space as usize);
+            let create_ix = create_account(
+                &ctx.accounts.payer.key,
+                &ctx.accounts.proof_buffer.key,
+                lamports,
+                space as u64,
+                &crate::id(),
+            );
+            invoke(
+                &create_ix,
+                &[
+                    ctx.accounts.payer.to_account_info(),
+                    ctx.accounts.proof_buffer.to_account_info(),
+                    ctx.accounts.system_program.to_account_info(),
+                ],
+            )?;
+            buffer[0..4].copy_from_slice(&max_depth.to_le_bytes());
+            buffer[4..8].copy_from_slice(&max_depth.to_le_bytes());
+            (max_depth, max_depth)
+        } else {
+            (
+                u32::from_le_bytes(
+                    buffer[0..4]
+                        .try_into()
+                        .map_err(|_| AccountCompressionError::InvalidProofBuffer)?,
+                ),
+                u32::from_le_bytes(
+                    buffer[4..8]
+                        .try_into()
+                        .map_err(|_| AccountCompressionError::InvalidProofBuffer)?,
+                ),
```

Comment:
Feels like a decent idea to just give this buffer a header & version it. In case people want to use u64s later.